### PR TITLE
Move interactive commands out of InteractiveFeatures

### PR DESCRIPTION
**Pull request by drognanar**
Moves the interactive commands out of interactive features in order to expose the QueryStatus for such commands.
Uses MEF to actually call command's implementation.

Puts the commands in dlls which get loaded by default when Roslyn starts.
This way interactive dlls load only when commands get executed and not when they get created.

### CODE REVIEW

**Comment by davkean** [comment]
tag @dotnet/roslyn-ide @jasonmalinowski

**Comment by drognanar** [comment]
FYI @dotnet/roslyn-interactive @KevinH-MS

File: src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            _executeInInteractiveHandlers = executeInInteractiveHandlers;
+        }
+
+        private IExecuteInInteractiveCommandHandler CommandHandler
+        {
+            get
+            {
+                return _executeInInteractiveHandlers
+                    .Where(handler => handler.Metadata.ContentTypes.Contains(ContentTypeNames.CSharpContentType))
+                    .SingleOrDefault().Value;
```
Comment (KevinH-MS) [inline src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs]: It would be good to cache this.  We don't want to execute the Linq query every time a command is executed.

File: src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+namespace Microsoft.CodeAnalysis.Editor.CSharp.Interactive
+{
+    /// <summary>
+    /// Implements a execute in interactive command handler.
+    /// </summary>
+    [ExportCommandHandler("Interactive Command Handler", ContentTypeNames.CSharpContentType)]
+    internal class CSharpExecuteInInteractiveCommandHandler
+        : ICommandHandler<ExecuteInInteractiveCommandArgs>
+    {
+        private readonly IEnumerable<Lazy<IExecuteInInteractiveCommandHandler, ContentTypeMetadata>> _executeInInteractiveHandlers;
```
Comment (KevinH-MS) [inline src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs]: Is ```IExecuteInInteraciveCommandHandler``` necessary, or can we just substitute ```ICommandHandler<ExecuteInInteractiveCommandArgs>``` everywhere?

File: src/EditorFeatures/Core/Extensibility/Commands/ExportInteractiveCommandAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    [MetadataAttribute]
+    [AttributeUsage(AttributeTargets.Class)]
+    internal class ExportInteractiveAttribute : ExportAttribute
```
Comment (KevinH-MS) [inline src/EditorFeatures/Core/Extensibility/Commands/ExportInteractiveCommandAttribute.cs]: Do we need a specialized type here?  I imagine there must be some similar implementation somewhere else in the Roslyn codebase.
(if so, it's fine to clean this up in a subsequent checkin)

File: src/VisualStudio/CSharp/Repl/Commands.vsct [hunk 4]
```diff
@@ -23,25 +23,9 @@
...
-        </Strings>
-      </Button>
     </Buttons>
   </Commands>
   
-  <VisibilityConstraints>
-    <VisibilityItem context="guidCSProjectContext" guid="guidCSharpInteractiveCommandSet" id="cmdidResetInteractiveFromProject" />
-    <VisibilityItem context="guidVBProjectContext" guid="guidCSharpInteractiveCommandSet" id="cmdidResetInteractiveFromProject" />
-  </VisibilityConstraints>
-
```
Comment (KevinH-MS) [inline src/VisualStudio/CSharp/Repl/Commands.vsct]: Where did these visibility constraints go?  Where they unnecessary?

File: src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            _executeInInteractiveHandlers = executeInInteractiveHandlers;
+        }
+
+        private IExecuteInInteractiveCommandHandler CommandHandler
+        {
+            get
+            {
+                return _executeInInteractiveHandlers
+                    .Where(handler => handler.Metadata.ContentTypes.Contains(ContentTypeNames.CSharpContentType))
```
Comment (KevinH-MS) [inline src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs]: Ideally, we would get an IContentType for these, and to IContentType.IsOfType (rather than a string comparison), because content types can "inherit" from one another, but this can be addressed in a subsequent checkin.

**Comment by KevinH-MS** [comment]
Other than caching the CommandHandler, LGTM :+1: 

The rest of my suggestions can be addressed in a subsequent checkin if necessary.

File: src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs (hunk 2 above)
Comment (drognanar) [inline src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs]: For now I made `IExecuteInInteraciveCommandHandler` in order to ensure that the exported handler definitely does not get picked up by the `CommandHandlerService` which imports all instances of `ICommandHandler`.

File: src/VisualStudio/CSharp/Repl/Commands.vsct (hunk 4 above)
Comment (drognanar) [inline src/VisualStudio/CSharp/Repl/Commands.vsct]: The visibility constraints have been completely replaced by the BeforeQueryStatus in the `AbstractResetInteractiveMenuCommand.cs`.
Constraints were used as a mechanism to defer loading of the reset command. However, it was important to actually call the QueryStatus directly in order to restrict the visibility of command further. Since QueryStatus is called directly constraints are no longer needed.

Comment (KevinH-MS) [inline src/VisualStudio/CSharp/Repl/Commands.vsct]: Gotcha

File: src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+using System.ComponentModel.Composition;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.Interactive
+{
+    /// <summary>
+    /// Implements a execute in interactive command handler.
+    /// </summary>
+    [ExportCommandHandler("Interactive Command Handler", ContentTypeNames.CSharpContentType)]
+    internal class CSharpExecuteInInteractiveCommandHandler
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs]: You probably should have many comments here explaining why we're doing this in this binary versus somewhere else.

File: src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        : ICommandHandler<ExecuteInInteractiveCommandArgs>
+    {
+        private readonly IEnumerable<Lazy<IExecuteInInteractiveCommandHandler, ContentTypeMetadata>> _executeInInteractiveHandlers;
+        private IExecuteInInteractiveCommandHandler _executeInInteractiveHandler;
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharp/Interactive/CSharpExecuteInInteractiveCommandHandler.cs]: Just use Lazy<T>?

File: src/VisualStudio/Core/Def/Commands.vsct [hunk 8]
```diff
@@ -339,6 +339,18 @@
           <CommandName>GoToImplementation</CommandName>
         </Strings>
       </Button>
+
+      <!-- Interactive context menu in solution explorer. -->
+      <Button guid="guidCSharpInteractiveCommandSet" id="cmdidResetInteractiveFromProject" priority="0x0100" type="Button">
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Commands.vsct]: Why a "C#" command being defined in Core/Def/Commands.vsct?

File: src/VisualStudio/Core/Def/Implementation/Interactive/AbstractResetInteractiveMenuCommand.cs [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                GetResetInteractiveFromProjectCommandID());
+
+            resetInteractiveFromProjectCommand.Supported = true;
+
+            resetInteractiveFromProjectCommand.BeforeQueryStatus += (_, __) =>
+            {
+                EnvDTE.Project project;
+                FrameworkName frameworkName;
+                GetActiveProject(out project, out frameworkName);
+                var targetFramework = project.Properties.Item("TargetFramework");
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/Interactive/AbstractResetInteractiveMenuCommand.cs]: What is using targetFramework?

File: src/VisualStudio/Core/Def/Implementation/Interactive/CSharpResetInteractiveMenuCommand.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using System.ComponentModel.Design;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Interactive
+{
+    internal sealed class CSharpResetInteractiveMenuCommand
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/Interactive/CSharpResetInteractiveMenuCommand.cs]: Why is this defined in the base assembly versus the C#/VB specific scripting assemblies?

**Comment by MattGertz** [comment]
Approved pending successful completion of tests.

**Comment by genlu** [comment]
retest vsi please

**Comment by genlu** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please.
@dotnet-bot retest prtest/win/dbg/unit64 please

**Comment by davkean** [comment]
These might continue to fail, given: https://github.com/dotnet/roslyn/issues/8860

**Comment by davkean** [comment]
Just merge this in given what I called out above.

**Comment by drognanar** [comment]
@davkean It looks like tests passed after all.

File: src/EditorFeatures/Core/CommandHandlers/IExecuteInInteractiveCommandHandler.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+using Microsoft.CodeAnalysis.Editor.Commands;
+
+namespace Microsoft.CodeAnalysis.Editor.CommandHandlers
+{
+    /// <summary>
+    /// Interface for the ExecuteInInteractiveCommand handler.
+    /// Ensures that the command handler can be exported via MEF
+    /// without actually being instantiated as all other command handlers.
+    /// </summary>
+    internal interface IExecuteInInteractiveCommandHandler
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/CommandHandlers/IExecuteInInteractiveCommandHandler.cs]: Why was this necessary? I'd expect that to work the same as everything else.
